transform-style:preserve-3d doesn't work across display:contents ancestors

mattwoodrow opened the pull request:
#### adf1f00fb1ff19e7927fb3f519fe19e7d99caf22
<pre>
transform-style:preserve-3d doesn&apos;t work across display:contents ancestors
<a href="https://bugs.webkit.org/show_bug.cgi?id=273627">https://bugs.webkit.org/show_bug.cgi?id=273627</a>
&lt;<a href="https://rdar.apple.com/127468969">rdar://127468969</a>&gt;

Reviewed by Alan Baradlay.

When looking for the ancestor element to see if it has &apos;preserve-3d&apos;, any intermediate
ancestors with display:contents should be skipped.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/transform3d-preserve3d-014-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/transform3d-preserve3d-014.html: Added.
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::flattenedParent):
(WebCore::RenderLayer::ancestorLayerIsDOMParent const):

Canonical link: <a href="https://commits.webkit.org/278499@main">https://commits.webkit.org/278499@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8e6277a1f260b0f4f6e361bec15ce80bac64ba2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50691 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29988 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3008 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53950 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1382 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52994 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36250 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1032 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41312 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52790 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27642 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43666 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22430 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/50542 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25017 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/922 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9132 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/44027 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46998 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/984 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55540 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/50194 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25793 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/882 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48726 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/50712 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27050 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/43793 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47787 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11122 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27918 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/57669 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26782 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11869 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4096,11 +4096,25 @@ Ref<HitTestingTransformState> RenderLayer::createLocalTransformState(RenderLayer
     if (!ancestor)
         return false;
-    if (renderer().element() && ancestor->renderer().element() == renderer().element()->parentElementInComposedTree())
+    auto parent = flattenedParent(renderer().element());
+    if (parent && ancestor->renderer().element() == parent)
```
I guess `flattenedParent(renderer().element()) == ancestor->renderer().element()` would be sufficient but ofc checking against parent is less error prone.

alanbaradlay replied:
(totally not suggesting to change it:)

mattwoodrow replied:
I guess I was worried about `nullptr == nullptr`, not sure how probable that actually is.

alanbaradlay replied:
yeah, probably close to 0, but then again being defensive here is not a bad thing.

mattwoodrow committed:
transform-style:preserve-3d doesn't work across display:contents ancestors
https://bugs.webkit.org/show_bug.cgi?id=273627
<rdar://127468969>

Reviewed by Alan Baradlay.

When looking for the ancestor element to see if it has 'preserve-3d', any intermediate
ancestors with display:contents should be skipped.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/transform3d-preserve3d-014-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/transform3d-preserve3d-014.html: Added.
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::flattenedParent):
(WebCore::RenderLayer::ancestorLayerIsDOMParent const):

Canonical link: https://commits.webkit.org/278499@main
